ddtrace/appsec/_asm_request_context.py

```diff
@@ -27,41 +32,107 @@
-_DD_EARLY_HEADERS_CONTEXTVAR = contextvars.ContextVar("datadog_early_headers_contextvar", default=None)
-_DD_EARLY_HEADERS_CASE_SENSITIVE_CONTEXTVAR = contextvars.ContextVar(
-    "datadog_early_headers_casesensitive_contextvar", default=False
-)
-_DD_BLOCK_REQUEST_CALLABLE = contextvars.ContextVar("datadog_block_request_callable_contextvar", default=None)
-_DD_WAF_CALLBACK = contextvars.ContextVar("datadog_early_waf_callback", default=None)
-_DD_WAF_RESULTS = contextvars.ContextVar("datadog_early_waf_results", default=([[], [], []]))
-_DD_WAF_SENT = contextvars.ContextVar("datadog_waf_adress_sent", default=None)
-_DD_IAST_TAINT_DICT = contextvars.ContextVar("datadog_iast_taint_dict", default={})
 
+class ASM_Environment:
+    def __init__(self, active=False):  # type: (bool) -> None
+        self.active = active
+        self.span = None
+        self.waf_addresses = {}  # type: dict[str, Any]
+        self.callbacks = {}  # type: dict[str, Any]
+        self.telemetry = {}  # type: dict[str, Any]
+        self.addresses_sent = set()  # type: set[str]
 
-def reset():  # type: () -> None
-    _DD_EARLY_IP_CONTEXTVAR.set(None)
-    _DD_EARLY_HEADERS_CONTEXTVAR.set(None)
-    _DD_EARLY_HEADERS_CASE_SENSITIVE_CONTEXTVAR.set(False)
-    _DD_BLOCK_REQUEST_CALLABLE.set(None)
-    _DD_WAF_SENT.set(set())
-    _DD_IAST_TAINT_DICT.set({})
 
+_ASM = contextvars.ContextVar("ASM_contextvar", default=ASM_Environment())
 
-def set_ip(ip):  # type: (Optional[str]) -> None
-    _DD_EARLY_IP_CONTEXTVAR.set(ip)
 
+def free_context_available():  # type: () -> bool
+    env = _ASM.get()
+    return env.active and env.span is None
 
-def get_ip():  # type: () -> Optional[str]
-    return _DD_EARLY_IP_CONTEXTVAR.get()
 
+def register(span):
+    env = _ASM.get()
+    if not env.active:
+        log.debug("registering a span with no active asm context")
+        return
+    env.span = span
+
+
+class _Data_handler:
+    main_id = 0
+
+    def __init__(self):
+        _Data_handler.main_id += 1
+        env = ASM_Environment(True)
+
+        self._id = _Data_handler.main_id
+        self.active = True
+        self.token = _ASM.set(env)
+
+        env.telemetry["waf_results"] = [], [], []
+        env.callbacks["context"] = []
+
+    def finalise(self):
+        if self.active:
+            env = _ASM.get()
+            # assert _CONTEXT_ID.get() == self._id
+            callbacks = env.callbacks.get("context", [])
+            for function in callbacks:
+                function()
+            _ASM.reset(self.token)
+            self.active = False
 
-def set_taint_dict(taint_dict):  # type: (dict) -> None
-    _DD_IAST_TAINT_DICT.set(taint_dict)
+
+def set_value(category, address, value):  # type: (str, str, Any) -> None
+    env = _ASM.get()
+    if not env.active:
+        log.debug("setting %s address %s with no active asm context", category, address)
+        return
+    dictionary = getattr(env, category, None)
```

Comment:
NIT: could we change this name? dictionary isn't a good description of the variable ;)